File: 04-linked-lists/01_reverse_list.py
# ============================================================
# PATTERN: LINKED LISTS
# ============================================================
# 1. **Two pointers (slow/fast)** -- cycle detection, find middle, nth from end
# 2. **Dummy head** -- simplifies edge cases when head might change
# 3. **Reverse** -- iterative (prev/curr/next) or recursive
# 4. **Merge** -- two-pointer merge like merge sort
#
# **C tips:** Use `struct ListNode { int val; struct ListNode* next; };`. Always `malloc` new nodes, `free` deleted nodes. Watch for dangling pointers.
#
# ---
# ============================================================

import sys, os; sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..'))
from test_runner import run_tests


class ListNode:
    def __init__(self, val=0, nxt=None):
        self.val = val
        self.next = nxt


def make_list(arr):
    dummy = ListNode(0)
    curr = dummy
    for v in arr:
        curr.next = ListNode(v)
        curr = curr.next
    return dummy.next


def to_array(head):
    arr = []
    while head:
        arr.append(head.val)
        head = head.next
    return arr


# Reverse Linked List (LC #206) -- Easy
# Reverse a singly linked list.
#
#   Key: Three pointers: prev, curr, next. Or recursive.

def reverse_list(head):
    """head is a ListNode. Return new head."""
    pass


def solve(arr):
    return to_array(reverse_list(make_list(arr)))


run_tests(solve, [
    (([1,2,3,4,5],),  [5,4,3,2,1]),
    (([1,2],),        [2,1]),
    (([1],),          [1]),
    (([],),           []),
])

File: 04-linked-lists/01_reverse_list.c
#include "../test_runner.h"

typedef struct ListNode {
    int val;
    struct ListNode* next;
} ListNode;

ListNode* make_node(int val) {
    ListNode* n = malloc(sizeof(ListNode));
    n->val = val; n->next = NULL;
    return n;
}

ListNode* make_list(int* arr, int n) {
    if (n == 0) return NULL;
    ListNode* head = make_node(arr[0]);
    ListNode* curr = head;
    for (int i = 1; i < n; i++) {
        curr->next = make_node(arr[i]);
        curr = curr->next;
    }
    return head;
}

void to_array(ListNode* head, int* out, int* len) {
    *len = 0;
    while (head) { out[(*len)++] = head->val; head = head->next; }
}

/* ---- WRITE YOUR SOLUTION HERE ---- */

ListNode* reverse_list(ListNode* head) {
    return NULL;
}

/* ---- END SOLUTION ---- */

int main() {
    int out[10], len;
    to_array(reverse_list(make_list(ARR(1,2,3,4,5), 5)), out, &len);
    TEST_ARR("5 nodes", out, ARR(5,4,3,2,1), 5);

    to_array(reverse_list(make_list(ARR(1,2), 2)), out, &len);
    TEST_ARR("2 nodes", out, ARR(2,1), 2);

    to_array(reverse_list(make_list(ARR(1), 1)), out, &len);
    TEST_ARR("1 node", out, ARR(1), 1);

    to_array(reverse_list(NULL), out, &len);
    TEST_INT("empty", len, 0);

    TEST_SUMMARY();
}
